KEP-1880 Multiple ServiceCIDR (#43469)
kubernetes/website: content/en/docs/tasks/network/extend-service-ip-ranges.md

---
reviewers:
- thockin
- dwinship
min-kubernetes-server-version: v1.29
title: Extend Service IP Ranges
content_type: task
---

<!-- overview -->
{{< feature-state state="alpha" for_k8s_version="v1.29" >}}

This document shares how to extend the existing Service IP range assigned to a cluster.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

{{< version-check >}}

<!-- steps -->

## API

> Kubernetes clusters that are created with the `MultiCIDRServiceAllocator`

Reviewer: Created? Can you turn this on after cluster creation?

Author: it is an option in the apiserver, so I will rephrase to make this more explicit,

> [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) will create a new ServiceCIDR object with the well-known name `kubernetes` from the kube-apiserver `--service-cluster-ip-range` flag values.

Reviewer: ```suggestion
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) will create a new ServiceCIDR object that takes the well-known name `kubernetes`, and that uses an IP address range
based on  the value of the `--service-cluster-ip-range` command line argument to kube-apiserver.
```

In the world of IT, flags are for boolean values only; the Golang ecosystem uses “flag” to mean all kinds of command line argument, but this confuses people whose main programming language is not Go.

Reviewer: Would be nice to avoid the misleading term “flag”.

Author: oh, I thought I applied the suggestion

Author: one sec

Author: hmm, this file has not the reference to flag, is in the other file

Reviewer: Seems sorted!

Author: changed all the occurrences of flag

> 

Reviewer: Do you need to enable any API groups for this to work? I'd assume that you do.

```sh
kubectl get servicecidr
```
```
NAME         CIDRS          AGE
kubernetes   10.96.0.0/28   17d
```

> The well-known `kubernetes` Service, that exposes the kube-apiserver endpoint to the Pods, uses as ClusterIP the first IP from the default ServiceCIDR.

Reviewer: ```suggestion
The well-known `kubernetes` Service, that exposes the kube-apiserver endpoint to the Pods, calculates
the first IP address from the default ServiceCIDR range and uses that IP address as its
cluster IP address.
```

```sh
kubectl get service kubernetes
```
```
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   17d
```

The default Service, in this case, uses the ClusterIP 10.96.0.1, that has the corresponding IPAddress object.

```sh
kubectl get ipaddress 10.96.0.1
```
```
NAME        PARENTREF
10.96.0.1   services/default/kubernetes
```

> The ServiceCIDRs are protected with finalizers, to avoid leaving Service ClusterIPs orphans,

Reviewer: ```suggestion
The ServiceCIDRs are protected with {{<glossary_tooltip text="finalizers" term_id="finalizer">}}, to avoid leaving Service ClusterIPs orphans;
```

the finalizer is only removed if there is another subnet that contains the existing IPAddresses or
there are no IPAddresses belonging to the subnet.

## Extend the number of available IPs for Services

There are cases that users will need to increase the number addresses available to Services, previously, increasing the Service range was a disruptive operation that could also cause data loss. With this new feature users only need to add a new ServiceCIDR to increase the number of available addresses.

> [![asciicast](https://asciinema.org/a/618594.svg)](https://asciinema.org/a/618594)

Reviewer: Please avoid referencing an external object here as they are hard to maintain and to localize.

You could use a YouTube video posted to a channel that the Kubernetes project controls, or you could skip it.

### Adding a new ServiceCIDR

> On a cluster with a 10.96.0.0/28 range for Services, there is only 2^(32-28) - 2 = 14 IP addresses avaalable. The `kubernetes.default`  Service is always created, that leaves us with only 13 possible Services.

Reviewer: ```suggestion
On a cluster with a 10.96.0.0/28 range for Services, there is only 2^(32-28) - 2 = 14 IP addresses available. The `kubernetes.default` Service is always created; you this example, that leaves you with only 13 possible Services.
```
We [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we) in docs for end users to read.

```sh
for i in $(seq 1 13); do kubectl create service clusterip "test-$i" --tcp 80 -o json | jq -r .spec.clusterIP; done
```
```
10.96.0.11
10.96.0.5
10.96.0.12
10.96.0.13
10.96.0.14
10.96.0.2
10.96.0.3
10.96.0.4
10.96.0.6
10.96.0.7
10.96.0.8
10.96.0.9
error: failed to create ClusterIP service: Internal error occurred: failed to allocate a serviceIP: range is full
```

> The number of available addresses can be increased by creating a new ServiceCIDR that extends or adds new IP ranges.

Reviewer: (nit; fine as-is for alpha)
```suggestion
You can increase the number of IP addresses available for Services, by creating a new ServiceCIDR
that extends or adds new IP address ranges.
```

```sh
cat <EOF | kubectl apply -f -
apiVersion: networking.k8s.io/v1alpha1
kind: ServiceCIDR
metadata:
  name: newcidr1
spec:
  cidrs:
  - 10.96.0.0/24
EOF
```
```
servicecidr.networking.k8s.io/newcidr1 created
```

> and this will allow us to create new Services with ClusterIPs that will be picked from this new range. 

Reviewer: ```suggestion
and this will allow you to create new Services with ClusterIPs that will be picked from this new range.
```

```sh
for i in $(seq 13 16); do kubectl create service clusterip "test-$i" --tcp 80 -o json | jq -r .spec.clusterIP; done
```
```
10.96.0.48
10.96.0.200
10.96.0.121
10.96.0.144
```

### Deleting a ServiceCIDR

> A ServiceCIDR can not be deleted if there are IPAddressess that depend on it

Reviewer: ```suggestion
You cannot delete a ServiceCIDR if there are IPAddresses that depend on the ServiceCIDR.
```

```sh
kubectl delete servicecidr newcidr1
```
```
servicecidr.networking.k8s.io "newcidr1" deleted
```

> this is managed by a controller that runs in the kube-controller-manager
> that participates on the deletion of the object, and removes
> the finalizer once it verifies it will not leave orphan IPAddresses.

Reviewer: ```suggestion
Kubernetes uses a finalizer on the ServiceCIDR to track this dependent relationship.
```

```sh
kubectl get servicecidr newcidr1 -o yaml
```
```
apiVersion: networking.k8s.io/v1alpha1
kind: ServiceCIDR
metadata:
  creationTimestamp: "2023-10-12T15:11:07Z"
  deletionGracePeriodSeconds: 0
  deletionTimestamp: "2023-10-12T15:12:45Z"
  finalizers:
  - networking.k8s.io/service-cidr-finalizer
  name: newcidr1
  resourceVersion: "1133"
  uid: 5ffd8afe-c78f-4e60-ae76-cec448a8af40
spec:
  cidrs:
  - 10.96.0.0/24
status:
  conditions:
  - lastTransitionTime: "2023-10-12T15:12:45Z"
    message: There are still IPAddresses referencing the ServiceCIDR, please remove
      them or create a new ServiceCIDR
    reason: OrphanIPAddress
    status: "False"
    type: Ready
```

By removing the Services containing the IP addresses that are blocking the deletion of the ServiceCIDR

```sh
for i in $(seq 13 16); do kubectl delete service "test-$i" ; done
```
```
service "test-13" deleted
service "test-14" deleted
service "test-15" deleted
service "test-16" deleted
```

> the controller will remove the finalizer and the ServiceCIDR will be finally deleted.

Reviewer: ```suggestion
the control plane notices the removal. The control plane then removes its finalizer,
so that the ServiceCIDR that was pending deletion will actually be removed.
```

the first IP address from the default ServiceCIDR range and uses that IP address as its
cluster IP address.

```sh
kubectl get service kubernetes
```
```
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   17d
```

The default Service, in this case, uses the ClusterIP 10.96.0.1, that has the corresponding IPAddress object.

```sh
kubectl get ipaddress 10.96.0.1
```
```
NAME        PARENTREF
10.96.0.1   services/default/kubernetes
```

The ServiceCIDRs are protected with {{<glossary_tooltip text="finalizers" term_id="finalizer">}}, to avoid leaving Service ClusterIPs orphans;
the finalizer is only removed if there is another subnet that contains the existing IPAddresses or
there are no IPAddresses belonging to the subnet.

## Extend the number of available IPs for Services

There are cases that users will need to increase the number addresses available to Services, previously, increasing the Service range was a disruptive operation that could also cause data loss. With this new feature users only need to add a new ServiceCIDR to increase the number of available addresses.

### Adding a new ServiceCIDR

> On a cluster with a 10.96.0.0/28 range for Services, there is only 2^(32-28) - 2 = 14 IP addresses available. The `kubernetes.default` Service is always created; you this example, that leaves you with only 13 possible Services.

Reviewer: Oops, my bad
```suggestion
On a cluster with a 10.96.0.0/28 range for Services, there is only 2^(32-28) - 2 = 14 IP addresses available. The `kubernetes.default` Service is always created; for this example, that leaves you with only 13 possible Services.
``

Author: done
